Add tabular data methods (#2879)
viamrobotics/docs: docs/build/program/apis/data-client.md

{{% /tab %}}
{{< /tabs >}}

### TabularDataBySQL

Obtain unified tabular data and metadata, queried with SQL.

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

> - organization_id [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.

Reviewer: ```suggestion
- `organization_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.
```
Parameter names need code styling

> - sql_query [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The SQL query to run.

Reviewer: ```suggestion
- `sql_query` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The SQL query to run.
```

**Returns:**

> - (`List[Dict[str, ValueTypes]]`): An array of data objects.

Reviewer: ```suggestion
- (List[Dict[str, ValueTypes]]): An array of data objects.
```

```python {class="line-numbers linkable-line-numbers"}
tabular_data = await data_client.tabular_data_by_sql(
    organization_id=organization_id,
    sql_query="SELECT * FROM readings"
)
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/data_client/index.html#viam.app.data_client.DataClient.tabular_data_by_sql).

{{% /tab %}}
{{< /tabs >}}

### TabularDataByMQL

Obtain unified tabular data and metadata, queried with MQL.

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

> - organization_id [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.

Reviewer: ```suggestion
- `organization_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.
```

> - mql_binary (List[bytes]): The MQL query to run as a list of BSON documents.

Reviewer: ```suggestion
- `mql_binary` (List[bytes]): The MQL query to run as a list of BSON documents.
```

**Returns:**

> - (`List[Dict[str, ValueTypes]]`): An array of data objects.

Reviewer: ```suggestion
- (List[Dict[str, ValueTypes]]): An array of data objects.
```

```python {class="line-numbers linkable-line-numbers"}

> tabular_data = await data_client.tabular_data_by_sql(

Reviewer: ```suggestion
tabular_data = await data_client.tabular_data_by_mql(
```
does code sample need `mql_binary`?

    organization_id=organization_id,

>     sql_query="SELECT * FROM readings"

Reviewer: ```suggestion
```

>     sql_query="SELECT * FROM readings"

Author: ```suggestion
    mql_query=[b"{ $match: {
            'robot_id': 'abcdef12-abcd-abcd-abcd-abcdef123456',
            'component_name': 'my-ultrasonic-sensor',
            'data.readings.distance': { $gt: .2 } } },
        { $count: 'numStanding' }"]
```

**Returns:**

- (List[Dict[str, ValueTypes]]): An array of data objects.

```python {class="line-numbers linkable-line-numbers"}
tabular_data = await data_client.tabular_data_by_sql(
    organization_id=organization_id,
    sql_query="SELECT * FROM readings LIMIT 5"
)
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/data_client/index.html#viam.app.data_client.DataClient.tabular_data_by_sql).

{{% /tab %}}
{{< /tabs >}}

### TabularDataByMQL

Obtain unified tabular data and metadata, queried with MQL.

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

- `organization_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.
- `mql_binary` (List[bytes]): The MQL query to run as a list of BSON documents.

**Returns:**

- (List[Dict[str, ValueTypes]]): An array of data objects.

```python {class="line-numbers linkable-line-numbers"}
tabular_data = await data_client.tabular_data_by_mql(
    organization_id=organization_id,
    mql_query=[
        bson.encode({'$match': {'location_id': '<location-id>'} }),
        bson.encode({ "$limit": 5 })
    ]
)
```

> For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/data_client/index.html#viam.app.data_client.DataClient.tabular_data_by_mql).

Author: ```suggestion
You must install the `pymongo` package and `import bson` to use the `tabular_data_by_mql` method.
For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/data_client/index.html#viam.app.data_client.DataClient.tabular_data_by_mql).
```

{{% /tab %}}
{{< /tabs >}}

### TabularDataBySQL

Obtain unified tabular data and metadata, queried with SQL.

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

- `organization_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.
- `sql_query` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The SQL query to run.

**Returns:**

- (List[Dict[str, ValueTypes]]): An array of data objects.

> ```python {class="line-numbers linkable-line-numbers"}

Author: need to test this

Author: will test when the next python sdk is released

tabular_data = await data_client.tabular_data_by_sql(
    organization_id=organization_id,
    sql_query="SELECT * FROM readings"
)
```

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/app/data_client/index.html#viam.app.data_client.DataClient.tabular_data_by_sql).

{{% /tab %}}
{{< /tabs >}}

### TabularDataByMQL

Obtain unified tabular data and metadata, queried with MQL.

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

- `organization_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization that owns the data.
- `mql_binary` (List[bytes]): The MQL query to run as a list of BSON documents.

**Returns:**

- (List[Dict[str, ValueTypes]]): An array of data objects.

> ```python {class="line-numbers linkable-line-numbers"}

Author: need to test this

tabular_data = await data_client.tabular_data_by_mql(
    organization_id=organization_id,
    mql_query=[

>         bson.encode({ '$match': { 'location_id': '<location-id>' } }),
>         bson.encode({ "$limit": 5 })

Author: ```suggestion
        bson.dumps({ '$match': { 'location_id': '<location-id>' } }),
        bson.dumps({ "$limit": 5 })
```

            'data.readings.distance': { $gt: .2 } } },

>         { $count: 'numStanding' }"]

Author: Hey @purplenicole730, I've been trying to figure out how I'd specify the mql query - I know this isn't it but I can't figure out how it would work - can you help?

Reviewer 2: I've never had to actually use mql queries either. I would suggest asking someone from the data team.

>     mql_query=[b"{ $match: {
>             'robot_id': 'abcdef12-abcd-abcd-abcd-abcdef123456',
>             'component_name': 'my-ultrasonic-sensor',
>             'data.readings.distance': { $gt: .2 } } },
>         { $count: 'numStanding' }"]

Author: ```suggestion
    mql_query=[
        bson.encode({'$match': {'location_id': '7kp7y4p393'} }),
        bson.encode({ "$limit": 5 })
    ]
```

```

> You must install the `pymongo` package and `import bson` to use the `tabular_data_by_mql` method.

Author: ```suggestion
You must `import bson` to create valid bson binary objects as input for the `tabular_data_by_mql` method.
```

{{% /tab %}}
{{< /tabs >}}

### ConfigureDatabaseUser

> Configure a database user for the Viam organization’s MongoDB Atlas Data Federation instance. It can also be used to reset the

Reviewer: ```suggestion
Configure a database user for the Viam organization’s MongoDB Atlas Data Federation instance.
```

> password of the existing database user.

Reviewer: ```suggestion
You can also use this to reset the password of the existing database user.
```
or something like that, more direct

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

> - organization_id [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization.

Reviewer: ```suggestion
- `organization_id` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The ID of the organization.
```

> - password [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The password of the user.

Reviewer: ```suggestion
- `password` [(str)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The password of the user.
```
